archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,264 @@
+end
+
+
+lemma radical_inequality {n : ℕ} (h : n ≥ 107) : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3) :=
+begin
+  rw sqrt_le_iff,
+  split,
+  { norm_num,
+    rw le_sqrt,
+    repeat {norm_cast, linarith}},
+  ring_exp,
+  rw [pow_two, ← sqrt_mul, sqrt_mul_self],
+  suffices : 24 * sqrt (1 + n) ≤ 2 * n + 36,
+  linarith,
+  rw mul_self_le_mul_self_iff,
+  ring_exp,
+  rw [pow_two, ← sqrt_mul, sqrt_mul_self],
+  --Not splitting into cases lead to a deterministic timeout on my machine
+  by_cases p: n ≥ 108,
+  norm_cast,
+  nlinarith,
+  have : n = 107 := by linarith,
+  subst this,
+  norm_num,
+  repeat {norm_cast, linarith},
+  repeat {norm_num, apply sqrt_nonneg},
+end
+
+/-We will later make use of the fact that there exists (l : ℕ) such that
+n ≤ 2 * l * l - 4 * l and 2 * l * l + 4 * l ≤ 2 * n for n ≥ 107.
+-/
+lemma exists_numbers_in_interval (n : ℕ) (hn : n ≥ 107): ∃ (l : ℤ), l ≥ 0 ∧
+((n : ℤ) + 4 * l ≤ 2 * l * l ∧ 2 * l * l + 4 * l ≤ 2 * n) :=
+begin
+  suffices : ∃ (l : ℤ), l ≥ 0 ∧ 2 * (l : ℝ) ≥ 2 + sqrt (4 + 2 * n) ∧ (l : ℝ) ≤ sqrt (1 + n) - 1,
+  cases this with l t,
+  use l,
+  refine ⟨t.1, ⟨lower_bound n l t.2.1, upper_bound n l t.1 t.2.2⟩⟩,
+  use floor (sqrt (1 + n) - 1),
+  split,
+  { simp only [ge_iff_le],
+    rw floor_nonneg,
+    norm_num,
+    rw le_sqrt,
+    repeat {norm_cast, linarith}},
+  refine ⟨_, floor_le (sqrt (1 + n) - 1)⟩,
+  { suffices : 2 * (sqrt (1 + n) - 2) ≥ 2 + sqrt (4 + 2 * n),
+    apply le_trans this _,
+    simp only [mul_le_mul_left, zero_lt_bit0, zero_lt_one],
+    suffices : sqrt (1 + n) - 1 ≤ ↑⌊sqrt (1 + n) - 1⌋ + 1,
+    linarith,
+    have t :  (⌈sqrt (1 + n) - 1⌉:ℝ) ≤ ⌊sqrt (1 + n) - 1⌋ + 1,
+    norm_cast,
+    exact ceil_le_floor_add_one _,
+    apply le_trans _ t,
+    exact le_ceil (sqrt (1 + ↑n) - 1),
+    suffices : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3),
+    linarith,
+    exact radical_inequality hn },
+end
+
+/-
+We show there exists a triplet a, b, c ∈ [n , 2n] with a < b < c and each of the sums (a + b),
+(b + c), (a + c) being a perfect square. Specifically, we consider the linear system of equations
+
+  a + b = (2 * l - 1) ^ 2
+  a + c = (2 * l) ^ 2
+  b + c = (2 * l + 1) ^ 2
+
+which can be solved to give
+
+  a = 2 * l * l - 4 * l
+  b = 2 * l * l + 1
+  c = 2 * l * l + 4 * l
+
+Therefore, it is enough to show that there exists a natural number l such that
+n ≤ 2 * l * l - 4 * l and 2 * l * l + 4 * l ≤ 2 * n for n ≥ 100.
+-/
+
+lemma exists_triplet_summing_to_squares (n : ℕ) (hn : n ≥ 100): (∃ (a b c : ℤ),
+(n : ℤ) ≤ a ∧ a < b ∧ b < c ∧ c ≤ 2 * n ∧  (∃ (k : ℤ), a + b = k * k) ∧
+(∃ (l : ℤ), a + c = l * l) ∧ (∃ (m : ℤ), b + c = m * m)) :=
+begin
+  --If n ≥ 107, we do not explicitly construct the triplet but use an existence
+  --argument from lemma above.
+  by_cases p : n ≥ 107,
+  have h := exists_numbers_in_interval n p,
+  cases h with l hl,
+  use (2 * l * l - 4 * l),
+  use (2 * l * l + 1),
+  use (2 * l * l + 4 * l),
+  refine ⟨_,⟨_,⟨_,⟨_,⟨_, ⟨_, _⟩⟩⟩⟩⟩⟩,
```

Comment:
`exists` is an inductive type, so you can set the `use` values in here too; I'll show you how below